DirectProgramming/DPC++FPGA/Tutorials/Features/device_global/README.md

```diff
@@ -0,0 +1,225 @@
+| Software                          | Intel&reg; oneAPI DPC++ Compiler <br> Intel&reg; FPGA Add-On for oneAPI Base Toolkit
+| What you will learn               | The basic usage of the `device_global` class <br> How initialize a `device_global` to non-zero values<br>
+| Time to complete                  | 15 minutes
+
+## Purpose
+This tutorial demonstrates a simple example of initializing a `device_global` to a non-zero value, and using it to keep state between multiple re-launches of a kernel.
+
+### Description of `device_global`
+`device_globals` are an extension that introduces device scoped memory allocations into SYCL that can be accessed within a kernel using syntax similar to C++ global variables, but that have unique instances per `sycl::device`. Similar to C++ global variables, `device_globals` have namespace scope and are visible to all kernels within that scope.
+
+`device_globals` can be templated on a list of compile-time-constant `properties` that can enable various optimization behaviors. One such `property` is `device_image_scope` which is used in the code sample. It limits the scope of a single instance of a `device_global` to a `device_image`, further details on this property can be found [here](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_device_global.asciidoc#properties-for-device-global-variables). The absence of the `device_image_scope` property is currently not supported by the compiler.
+
+`device_globals` can be used to store state across multiple re-launches of a kernel without having to pass in a `buffer` as a kernel argument. This is useful in cases where components are nodes in a state-machine.
+
+### Non-zero-initialization of a `device_global` - pre C++20
+Without C++20 `consteval` keyword support, `device_globals` are zero-initialized. If there is a known first usage of a `device_global` in device code, there exists a technique, described below, that allows a `device_global` to be initialized to a non-zero value. 
+
+Using a second `device_global<bool>` representing a flag that will control when to initialize the `device_global` to a non-zero value. This flag will get zero-initialized to `false`. Once initialization happens, the flag is set to `true` and initialization code doesn't execute on subsequent relaunches of the kernel.
+
+```cpp
+using namespace sycl;
+using namespace sycl::ext::oneapi::experimental;
+
+device_global<int> val;
+device_global<bool> is_val_initialized;
+
+int main () {
+  queue Q;
+  Q.submit([&](sycl::handler& h) {
+    h.single_task([=] {
+      // Initialization happens only once
+      if (!is_val_initialized) {
+        // do initialization
+        is_val_initialized.get() = true;
+      }
+
+      // uses of `val`
+    });
+  });
+}
+```
+
+
+### Non-zero-initialization of a `device_global`- with C++20
```

Comment:
Yeah i was wrestling with that when i was writing this. Ultimately i decided that it would be really natural for user to want constant value-initialization on device_globals - and i wanted to preemptively stop feature requests and explain that it is in the works and is gated by a modern C++20 feature that is not fully implemented by any compiler.

Do you still think i should cut this?

Reply:
I think that in that case you should just have a quick mention as a note in a "Use model" section saying something like what I wrote above: 
"note that the technique below for initializing the device_global variables will be simplified once we support C++20 to device_global<int> a {42};"